### PR TITLE
Bombman frequently start 

### PR DESCRIPTION
**Pull request by Alx-Lai**

### CODE REVIEW

File: extensions/bombman/standalone.mjs [hunk 1]
```diff
@@ -80,6 +89,46 @@ class Standalone {
...
+   * start Game
+   * @return {Boolean} success or not
+   */
+  async startGame() {
+    // TODO:start Game
+    if (this.gameStarted) {
+      await this.resetGame();
+    }
+    if (this.playerIDs.size < 2) {
+      console.log('[bombman] player not enough to start (< 3)');
```
Comment (john0312) [inline extensions/bombman/standalone.mjs:124]: Consider emitting a notification to all users participating. Search for `showNotification` and `showAnnouncement` for examples.
```
this.helper.callS2cAPI(pid, 'notification', 'showNotification', 5000, `...`);
```

File: extensions/bombman/standalone.mjs [hunk 2]
```diff
@@ -80,6 +89,46 @@ class Standalone {
...
+      return false;
+    }
+    for (const playerID of this.playerIDs) {
+      this.teleportPlayer(playerID);
```
Comment (john0312) [inline extensions/bombman/standalone.mjs:128]: Consider showing notification or announcements, as above.

File: extensions/bombman/standalone.mjs [hunk 3]
```diff
@@ -250,13 +294,9 @@ class Standalone {
...
    */
-  async s2s_sf_startBombman(srcExt, playerID, kwargs, sfInfo) {
+  async s2s_sf_joinBombman(srcExt, playerID, kwargs, sfInfo) {
     const {next} = kwargs;
-    if (this.gameStarted) {
-      console.log('[bombman] The game has already started');
-      return next;
-    }
-    this.gameStarted = true;
+    this.playerIDs.add(playerID);
```
Comment (john0312) [inline extensions/bombman/standalone.mjs:299]: Ditto on notifications.

File: extensions/bombman/standalone.mjs [hunk 4]
```diff
@@ -268,40 +308,10 @@ class Standalone {
...
-              name: LAYER_BOMB_EXPLODE.layerName,
-              cells: Array.from(this.explodeCells.values()),
-            }),
-        );
-      }
-    } catch (e) {
-      console.error(e);
-      return FSM_ERROR;
+    if (this.playerIDs.has(playerID)) {
+      this.playerIDs.delete(playerID);
```
Comment (john0312) [inline extensions/bombman/standalone.mjs:314]: Do we teleport them if the game has already started?

File: extensions/bombman/standalone.mjs [hunk 5]
```diff
@@ -315,6 +325,74 @@ class Standalone {
...
+    target.x = 10;
+    target.y = 5;
+    this.helper.teleport(player.playerID, target, true);
+  }
+
+  async terminateGame() {
+    for (const playerID of this.playerIDs) {
+      console.log(playerID, 'Win');
+      // TODO: give item
+      // this.helper.callS2sAPI('item', 'giveItem', playerID, 'bombman_trophy', 1);
```
Comment (john0312) [inline extensions/bombman/standalone.mjs:360]: Consider making the trophy item configurable through the configuration. If the configuration is not set, then we can skip.

File: extensions/bombman/standalone.mjs [hunk 6]
```diff
@@ -80,6 +89,46 @@ class Standalone {
 
     // set cooldown
     this.cooldown = new Set(); // TODO: Cooldown Manager
+
+    this.playerIDs = new Set(); // store playerID
```
Comment (ktpss95112) [inline extensions/bombman/standalone.mjs:93]: Consider a more descriptive naming.

File: extensions/bombman/standalone.mjs [hunk 7]
```diff
@@ -80,6 +89,46 @@ class Standalone {
...
+  setGameInterval() {
+    alarm.recurring(START_GAME_INTERVAL, this.startGame.bind(this));
+  }
+
+  /**
+   * start Game
+   * @return {Boolean} success or not
+   */
+  async startGame() {
+    // TODO:start Game
```
Comment (ktpss95112) [inline extensions/bombman/standalone.mjs:119]: redundant?

File: extensions/bombman/standalone.mjs (hunk 2 above)
Comment (ktpss95112) [inline extensions/bombman/standalone.mjs:128]: await

File: extensions/bombman/standalone.mjs [hunk 8]
```diff
@@ -93,7 +142,7 @@ class Standalone {
         const width = cell.w ?? 1;
         const height = cell.h ?? 1;
         if (cell.x <= cellCoord.x && cellCoord.x < cell.x + width &&
-            cell.y <= cellCoord.y && cellCoord.y < cell.y + height) {
+          cell.y <= cellCoord.y && cellCoord.y < cell.y + height) {
```
Comment (ktpss95112) [inline extensions/bombman/standalone.mjs:145]: eslint would tell you how many space you should use

File: extensions/bombman/standalone.mjs (hunk 4 above)
Comment (Alx-Lai) [reply extensions/bombman/standalone.mjs:314]: nope 因為我上次看bombman 的地圖是封閉的 所以我猜他們touch 不到這個npc

File: extensions/bombman/standalone.mjs [hunk 9]
```diff
@@ -315,6 +325,74 @@ class Standalone {
   async s2s_provideStateFunc(srcExt, registerFunc) {
     this.helper.callS2sAPI(srcExt, registerFunc, this.helper.getListOfStateFunctions(this));
   }
+
+  /**
+   * kill player and tp
+   * @param {String} playerID
```
Comment (ktpss95112) [inline extensions/bombman/standalone.mjs:331]: missing parameter `endGame` (optional)

File: extensions/bombman/standalone.mjs [hunk 10]
```diff
@@ -315,6 +325,74 @@ class Standalone {
...
+   */
+  async killPlayer(playerID, endGame = false) {
+    const player = this.helper.gameState.getPlayer(playerID);
+    const target = player.mapCoord.copy();
+    target.x = 1;
+    target.y = 1;
```
Comment (ktpss95112) [inline extensions/bombman/standalone.mjs:337]: Is this intended to be hardcoded?

File: extensions/bombman/standalone.mjs [hunk 11]
```diff
@@ -315,6 +325,74 @@ class Standalone {
...
+    this.helper.teleport(playerID, target, true);
+    this.playerIDs.delete(playerID);
+    if (this.playerIDs.size <= 1 && !endGame) {
+      await this.terminateGame();
+    }
+  }
+
+  /**
+   * @param {String} playerID playerID
```
Comment (ktpss95112) [inline extensions/bombman/standalone.mjs:346]: `@param {String} playerID - playerID`

File: extensions/bombman/standalone.mjs [hunk 12]
```diff
@@ -315,6 +325,74 @@ class Standalone {
...
+   */
+  teleportPlayer(playerID) {
+    const player = this.helper.gameState.getPlayer(playerID);
+    const target = player.mapCoord.copy();
+    target.x = 10;
+    target.y = 5;
```
Comment (ktpss95112) [inline extensions/bombman/standalone.mjs:345]: Is this intended to be hardcoded?

File: extensions/bombman/standalone.mjs [hunk 13]
```diff
@@ -315,6 +325,74 @@ class Standalone {
...
+    this.helper.teleport(player.playerID, target, true);
+  }
+
+  async terminateGame() {
+    for (const playerID of this.playerIDs) {
+      console.log(playerID, 'Win');
```
Comment (ktpss95112) [inline extensions/bombman/standalone.mjs:354]: Should be commented?

File: extensions/bombman/standalone.mjs [hunk 14]
```diff
@@ -315,6 +325,74 @@ class Standalone {
...
+      // TODO: give item
+      // this.helper.callS2sAPI('item', 'giveItem', playerID, 'bombman_trophy', 1);
+    }
+    await this.resetGame();
+
+    for (const playerID of this.playerIDs) {
+      this.killPlayer(playerID, true);
```
Comment (ktpss95112) [inline extensions/bombman/standalone.mjs:361]: await

**Review by ktpss95112 — COMMENTED** [review]
as the comments show

**Review by ktpss95112 — APPROVED** [review]

**Pull request merged**